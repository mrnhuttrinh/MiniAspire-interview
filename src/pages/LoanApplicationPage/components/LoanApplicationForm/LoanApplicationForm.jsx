import React from 'react';
import block from 'bem-cn';
import { compose } from 'redux';
import { connect } from 'react-redux';
import {
  Field, reduxForm, getFormValues,
} from 'redux-form';

import { InputField, SelectField } from '../../../../shared/reduxForm';
import {
  required, email, phoneNumber,
  minNumber, maxNumber,
} from '../../../../shared/validators';

import './LoanApplicationForm.scss';

const b = block('loan-application-form');

const zeroNumber = minNumber(0);
const billionNumber = maxNumber(1000000000);

function LoanApplicationForm(props) {
  const { handleSubmit, updateLoanForm } = props;

  const handleFormSubmit = (values) => {
    updateLoanForm(values);
  };

  return (
    <form className={b()} onSubmit={handleSubmit(handleFormSubmit)}>
      <div className="title">Contact Information</div>
      <Field
        validate={[required]}
        options={[
          { key: 'Mr', text: 'Mr'},
          { key: 'Mrs', text: 'Mrs'},
          { key: 'Ms', text: 'Ms'},
        ]}
        component={SelectField}
        name="title"
        label="Title"
      />

      <Field
        validate={[required]}
        component={InputField}
        name="firstName"
        label="First Name"
      />

      <Field
        validate={[required]}
        component={InputField}
        name="lastName"
        label="Last Name"
      />

      <Field
        validate={[required]}
        component={InputField}
        name="dateOfBirth"
        label="Date Of Birth"
        type="date"
      />

      <Field
        validate={[required]}
        options={[
          { key: 'Male', text: 'Male'},
          { key: 'Female', text: 'Female'},
          { key: 'Other', text: 'Other'},
        ]}
        component={SelectField}
        name="material"
        label="Material"
      />

      <Field
        validate={[required, phoneNumber]}
        component={InputField}
        name="phone"
        label="Phone"
      />
      
      <Field
        component={InputField}
        validate={[required, email]}
        name="email"
        label="Email"
      />

      <Field
        component={InputField}
        validate={[required]}
        name="address"
        label="Address"
      />

      <br />
      <hr />
      <div className="title">Loan Information</div>
      <Field
        component={InputField}
        validate={[required, zeroNumber, billionNumber]}
        name="amount"
        label="Amount"
        type="number"
      />

      <Field
        validate={[required]}
        options={[
          { key: 'Weekly', text: 'Weekly'},
          { key: 'Monthly', text: 'Monthly'},
          { key: 'Quarterly', text: 'Quarterly'},
          { key: 'Annually', text: 'Annually'},
        ]}
        component={SelectField}
        name="term"
        label="Term"
      />

      <div>
        <button type="submit" className="submit-button">
          Send Application Now
        </button>
      </div>
    </form>
  );
}

const LOAN_APPLICATION_FORM = 'loanApplicationForm';

const mapStateToProps = (state) => {
  return {
    values: getFormValues(LOAN_APPLICATION_FORM)(state)
  };
};

export { LoanApplicationForm };

export default compose(
  reduxForm({
    form: LOAN_APPLICATION_FORM,
    initialValues: {
      title: 'Mr',
      material: 'Male',
      term: 'Weekly'
    },
  }),
  connect(mapStateToProps),
)(LoanApplicationForm);
