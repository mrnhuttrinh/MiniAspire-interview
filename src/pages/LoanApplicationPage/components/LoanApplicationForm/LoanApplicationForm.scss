.loan-application-form {
  margin: 0 auto 30px;

  width: 600px;
  height: auto;
  background-color: rgba($color: #000000, $alpha: 0.05);
  border-radius: 5px;
  box-shadow: 0 0 5px 0;
  padding: 20px;
  display: flex;
  flex-direction: column;

  .submit-button {
    padding: 5px 15px;
    margin-left: 50%;
    margin-top: 20px;
    transform: translateX(-50%);
    border-radius: 5px;
  }

  .title {
    margin: 10px 0;
    font-weight: bold;
    font-size: 18px;
  }
}