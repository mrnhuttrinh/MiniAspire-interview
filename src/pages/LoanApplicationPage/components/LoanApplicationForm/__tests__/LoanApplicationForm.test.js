import React from 'react';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { LoanApplicationForm } from '../LoanApplicationForm';
import {default as LoanApplicationFormConnected } from '../LoanApplicationForm';

const buildStore = configureStore();

describe('LoanApplicationForm', () => {
  it('should render LoanApplicationForm', () => {
    const handleSubmit = jest.fn();
    const wrapper = shallow(
      <LoanApplicationForm handleSubmit={handleSubmit} />
    );
    expect(wrapper).toMatchSnapshot();
    expect(handleSubmit).toHaveBeenCalled();
  });

  it('should render LoanApplicationForm with hoc: redux-form and connect', () => {
    const initialState = { loanApplication: null };
    const store = buildStore(initialState);
    const updateLoanForm = jest.fn();
    const wrapper = mount(
      <Provider store={store}>
        <LoanApplicationFormConnected updateLoanForm={updateLoanForm} />
      </Provider>
    );

    const handleSubmit = jest.fn();
    wrapper.find('form').simulate('submit', {
      submit: handleSubmit,
    });

    expect(updateLoanForm).toHaveBeenCalled();
  });
});
