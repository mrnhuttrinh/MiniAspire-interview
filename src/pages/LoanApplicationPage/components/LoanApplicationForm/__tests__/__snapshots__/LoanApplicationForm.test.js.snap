// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoanApplicationForm should render LoanApplicationForm 1`] = `
<form
  className="loan-application-form"
>
  <div
    className="title"
  >
    Contact Information
  </div>
  <Field
    component={[Function]}
    label="Title"
    name="title"
    options={
      Array [
        Object {
          "key": "Mr",
          "text": "Mr",
        },
        Object {
          "key": "Mrs",
          "text": "Mrs",
        },
        Object {
          "key": "Ms",
          "text": "Ms",
        },
      ]
    }
    validate={
      Array [
        [Function],
      ]
    }
  />
  <Field
    component={[Function]}
    label="First Name"
    name="firstName"
    validate={
      Array [
        [Function],
      ]
    }
  />
  <Field
    component={[Function]}
    label="Last Name"
    name="lastName"
    validate={
      Array [
        [Function],
      ]
    }
  />
  <Field
    component={[Function]}
    label="Date Of Birth"
    name="dateOfBirth"
    type="date"
    validate={
      Array [
        [Function],
      ]
    }
  />
  <Field
    component={[Function]}
    label="Material"
    name="material"
    options={
      Array [
        Object {
          "key": "Male",
          "text": "Male",
        },
        Object {
          "key": "Female",
          "text": "Female",
        },
        Object {
          "key": "Other",
          "text": "Other",
        },
      ]
    }
    validate={
      Array [
        [Function],
      ]
    }
  />
  <Field
    component={[Function]}
    label="Phone"
    name="phone"
    validate={
      Array [
        [Function],
        [Function],
      ]
    }
  />
  <Field
    component={[Function]}
    label="Email"
    name="email"
    validate={
      Array [
        [Function],
        [Function],
      ]
    }
  />
  <Field
    component={[Function]}
    label="Address"
    name="address"
    validate={
      Array [
        [Function],
      ]
    }
  />
  <br />
  <hr />
  <div
    className="title"
  >
    Loan Information
  </div>
  <Field
    component={[Function]}
    label="Amount"
    name="amount"
    type="number"
    validate={
      Array [
        [Function],
        [Function],
        [Function],
      ]
    }
  />
  <Field
    component={[Function]}
    label="Term"
    name="term"
    options={
      Array [
        Object {
          "key": "Weekly",
          "text": "Weekly",
        },
        Object {
          "key": "Monthly",
          "text": "Monthly",
        },
        Object {
          "key": "Quarterly",
          "text": "Quarterly",
        },
        Object {
          "key": "Annually",
          "text": "Annually",
        },
      ]
    }
    validate={
      Array [
        [Function],
      ]
    }
  />
  <div>
    <button
      className="submit-button"
      type="submit"
    >
      Send Application Now
    </button>
  </div>
</form>
`;
